board/safety/safety_volkswagen.h

```diff
@@ -0,0 +1,162 @@
+    int sign = (GET_BYTE(to_push, 6) & 0x80) >> 7;
+    if (sign == 1) {
+      torque_driver_new *= -1;
+    }
+
+    update_sample(&vw_torque_driver, torque_driver_new);
+  }
+
+  // Monitor ACC_06.ACC_Status_ACC for stock ACC status. Because the current MQB port is lateral-only, OP's control
+  // allowed state is directly driven by stock ACC engagement. Permit the ACC message to come from either bus, in
+  // order to accommodate future camera-side integrations if needed.
+  if (addr == MSG_ACC_06) {
+    int acc_status = (GET_BYTE(to_push,7) & 0x70) >> 4;
+    controls_allowed = ((acc_status == 3) || (acc_status == 4) || (acc_status == 5)) ? 1 : 0;
+  }
+}
+
+static int volkswagen_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
+  int addr = GET_ADDR(to_send);
+  int tx = 1;
+
+  // Safety check for HCA_01 Heading Control Assist torque.
+  if (addr == MSG_HCA_01) {
+    bool violation = false;
+
+    int desired_torque = GET_BYTE(to_send, 2) | ((GET_BYTE(to_send, 3) & 0x3F) << 8);
+    int sign = (GET_BYTE(to_send, 3) & 0x80) >> 7;
+    if (sign == 1) {
+      desired_torque *= -1;
+    }
+
+    uint32_t ts = TIM2->CNT;
+
+    if (controls_allowed) {
+
+      // *** global torque limit check ***
+      violation |= max_limit_check(desired_torque, VW_MAX_STEER, -VW_MAX_STEER);
+
+      // *** torque rate limit check ***
+      violation |= driver_limit_check(desired_torque, vw_desired_torque_last, &vw_torque_driver,
+        VW_MAX_STEER, VW_MAX_RATE_UP, VW_MAX_RATE_DOWN,
+        VW_DRIVER_TORQUE_ALLOWANCE, VW_DRIVER_TORQUE_FACTOR);
+      vw_desired_torque_last = desired_torque;
+
+      // *** torque real time rate limit check ***
+      violation |= rt_rate_limit_check(desired_torque, vw_rt_torque_last, VW_MAX_RT_DELTA);
+
+      // every RT_INTERVAL set the new limits
+      uint32_t ts_elapsed = get_ts_elapsed(ts, vw_ts_last);
+      if (ts_elapsed > VW_RT_INTERVAL) {
+        vw_rt_torque_last = desired_torque;
+        vw_ts_last = ts;
+      }
+    }
+
+    // no torque if controls is not allowed
+    if (!controls_allowed && (desired_torque != 0)) {
+      violation = true;
+    }
+
+    // reset to 0 if either controls is not allowed or there's a violation
+    if (violation || !controls_allowed) {
+      vw_desired_torque_last = 0;
+      vw_rt_torque_last = 0;
+      vw_ts_last = ts;
+    }
+
+    if (violation) {
+      tx = 0;
+    }
+  }
+
+  // 1 allows the message through
+  return tx;
+}
+
+static int volkswagen_fwd_hook(int bus_num, CAN_FIFOMailBox_TypeDef *to_fwd) {
+  int addr = GET_ADDR(to_fwd);
+  int bus_fwd = -1;
+
+  // NOTE: Will need refactoring for other bus layouts, such as no-forwarding at camera or J533 running-gear CAN
+
+  switch(bus_num) {
+    case 0:
+      if(addr == MSG_GRA_ACC_01) {
+        // OP intercepts, filters, and updates the cruise-control button messages before they reach the ACC radar.
+        bus_fwd = -1;
```

Comment:
mmm... this is the only thing I'm not sure about. For example, what happens if you mistakenly send a set press to the radar? Does the radar start commanding acceleration?

Reply:
A fair question, and the answer would be yes. That said, I'm not sure if we have more exposure than any other port that's set up to spam cancel to comply with Comma strict safety, or spam resume to get stock ACC moving again in traffic. The only thing special about this port is the message is CRC-and-counter protected, so we have to stick ourselves in the middle of the message stream rather than just blat things out on the wire. Thoughts welcome.

Reply:
well, other ports have a check to make sure that only the cancel button can be sent.
I think we need that here as well. See https://github.com/commaai/panda/blob/master/board/safety/safety_honda.h#L156

Anyway, happy to merge this PR as is. It's good enough.

Reply:
> I'm also curious about why this is needed at all: I'm still not convinced but I'll know more when I see the port to openpilot.

A preview: https://github.com/jyoung8607/openpilot/blob/vw-064-devel/selfdrive/car/volkswagen/carcontroller.py#L98-L142

The short version: I want to send cancel and resume at a minimum, and I expect I would piss off the radar if it started getting duplicate/out-of-order counter values, although I admit I never tried. I also want to actually STOP set/resume from getting through if OP is not engageable, and for that I definitely have to get in the middle.

> I think we need something like that here as well. See https://github.com/commaai/panda/blob/master/board/safety/safety_honda.h#L156
> Anyway, happy to merge this PR as is. It's good enough.
 
Hmm, I like that. It's valid to restrict OP from sending resume unless stock ACC is already engaged, which it will be in stop-and-go traffic where OP needs to do so and it wouldn't be surprising to the user for the car to move. If that condition isn't met, I could potentially have Panda safety require any set/resume button press to be received from the car on 0 before it can be transmitted to the radar on 2.

Per our conversation, I'll take that as a TODO item, and try it out sometime after the current PR is merged.